crates/storage/storage-api/src/chain.rs

```diff
@@ -0,0 +1,71 @@
+use crate::{
+    BlockNumReader, BlockReader, DBProvider, HeaderProvider, TransactionsProvider,
+    WithdrawalsProvider,
+};
+use reth_db_api::transaction::{DbTx, DbTxMut};
+use reth_primitives::BlockHashOrNumber;
+use reth_primitives_traits::NodePrimitives;
+use reth_storage_errors::provider::ProviderResult;
+use std::fmt::Debug;
+
+/// Trait that implements how complex types (eg. Block) should be read or written to disk.
+pub trait ChainStorageReader: Send + Sync + Unpin + Default + Debug {
+    /// Primitive types of the node.
+    type Primitives: NodePrimitives;
+
+    /// Returns the block with given id from storage.
+    ///
+    /// Returns `None` if block is not found.
+    fn read_block<P>(
+        &self,
+        provider: &P,
+        id: BlockHashOrNumber,
+    ) -> ProviderResult<Option<<Self::Primitives as NodePrimitives>::Block>>
+    where
+        P: DBProvider<Tx: DbTx>
+            + TransactionsProvider
+            + BlockReader
+            + WithdrawalsProvider
+            + HeaderProvider
+            + BlockNumReader;
```

Comment:
wondering if we can avoid those bounds on trait level? I guess we can make the entire trait generic over the provider?

Reply:
yeah these are problematic. 
I believe @joshieDo also discussed changing to the just to the actual dbprovider instance so we don't need to enforce these traits on the storage trait itself